aiida/backends/sqlalchemy/utils.py

```diff
@@ -533,3 +533,20 @@ def alembic_command(selected_command, *args, **kwargs):
             al_command(alembic_cfg, message=args[0][0])
         else:
             al_command(alembic_cfg, *args, **kwargs)
+
+
+def delete_nodes_and_connections_sqla(pks_to_delete):
+    """
+    Delete all nodes corresponding to pks in the input.
+    :param pks_to_delete: A list, tuple or set of pks that should be deleted.
+    """
+    from aiida.backends import sqlalchemy as sa
+    from aiida.backends.sqlalchemy.models.node import DbNode, DbLink
+
+    session = sa.get_scoped_session()
+
+    with session.begin(subtransactions=True):
```

Comment:
At aiida.orm.implementation.sqlalchemy.code.delete_code I see the following 

```
        from aiida.backends.sqlalchemy import get_scoped_session
        session = get_scoped_session()
        session.begin(subtransactions=True)
        try:
            code.dbnode.delete()
            session.commit()
        except:
            session.rollback()
            raise
```

Which should be the verbose way of what you wrote. I suppose that at the end of the with clause/statement, there is an auto-commit (according to http://docs.sqlalchemy.org/en/latest/orm/session_transaction.html#session-autocommit).

It is worth verifying that for any kind of error in the queries, there is a rollback.I suppose there is but...
The same for Django

Reply:
The django part is taken from the existing script, and the same logic is replicated for sqlalchemy.
The django code is working on big databases, I was using it in a while in the current form. 
Regarding your comment on verbosity, I disagree: I do explicit deletion of the links, since it's more robust than relying on deletion cascades (which I'm not sure is implemented) for the links. Maybe above example for code deletion should be checked?

From the docs that you cite I understand that using the session in a try-except clause or with a context-manager is equivalent. As for the 'it is worth verifying': That's the reviewers job ;)

Reply:
It is not actually a proof of correctness that it runs on big databases (it's an indication). I also never mentioned to do deletion propagation etc.

An overall comment. The reviewers' job is to make some constructive comments on the code and highlight some potential corner cases (of course they should be somehow answered constructively since the author asked the reviewer's opinion & the reviewer invested some time to give feedback). 

Unfortunately, I don't have the time to check your code for all these corner cases.

Reply:
Me see problem, I make fix... Seriously, let's not have a big discussion here, about proofs of correctness, and the role of reviewers. I added  the explicit try-and-except clause that you wanted and made some more tests with deliberately wrong queries. The rollback works. Actually, I found that, different from Django, the cascading is not implemented for group-membership, so I also delete on the dbnode_dbgroup table  in 83b0478 .

Reply:
Thanks for the changes